File: model/src/main/java/net/qutester/model/step/StepDef.kt
package net.qutester.model.step

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import net.qutester.model.enums.StepPhaseEnum
import net.qutester.model.infrastructure.path.Path
import net.qutester.model.text.StepPattern


@JsonIgnoreProperties(ignoreUnknown = true)
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "@type")
@JsonSubTypes(value = [
        JsonSubTypes.Type(value = BasicStepDef::class, name = "BASIC_STEP"),
        JsonSubTypes.Type(value = ComposedStepDef::class, name = "COMPOSED_STEP"),
        JsonSubTypes.Type(value = UndefinedStepDef::class, name = "UNDEFINED_STEP")
])
interface StepDef {
    val id: String
    val path: Path
    val phase: StepPhaseEnum
    val stepPattern: StepPattern //TODO: rename to StepSignature
    val description: String?
    val tags: List<String>

    @JsonIgnore
    fun getText(): String {
        var result: String = phase.name;
        result += " ";

        result += stepPattern.getAsText();

        return result;
    }
}

File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/testerum-model/model/step/def/step-def.ts
import {Path} from "../../path";
import {StepPhaseEnum} from "./step-phase-enum";
import {StepPattern} from "./pattern/step-pattern";

export interface StepDef {
    id: string;
    path: Path;
    phase: StepPhaseEnum;
    stepPattern: StepPattern;
    description: string|null;
    tags: Array<string>;
}

export enum StepDefType {
    BASIC_STEP,
    COMPOSED_STEP,
    UNDEFINED_STEP,
}
